### PR TITLE
2D Renderer Code Refactor

### PR DESCRIPTION
**Pull request by UnityPaul**
### Purpose of this PR
2D Renderer code refactor.

---
### Testing status

**Manual Tests**: What did you do?
- [x] Opened test project + Run graphic tests locally
- [x] Built a player
- [ ] Checked new UI names with UX convention
- [ ] Tested UI multi-edition + Undo/Redo + Prefab overrides + Alignment in Preset
- [ ] C# and shader warnings (supress shader cache to see them)
- [ ] Checked new resources path for the reloader (in developer mode, you have a button at end of resources that check the paths)

**Automated Tests**:  A few runtime tests are removed due to code structure changes.

---
### Comments to reviewers
N/A

### CODE REVIEW

File: com.unity.render-pipelines.universal/Editor/2D/Light2DEditor.cs [hunk 1]
```diff
@@ -633,7 +639,11 @@ public override void OnInspectorGUI()
             }
 
             AnalyticsTrackChanges(serializedObject);
-            serializedObject.ApplyModifiedProperties();
+            if (serializedObject.ApplyModifiedProperties())
+            {
+                if(meshChanged)
+                    lightObject.UpdateMesh();
+            }
```
Comment (rainsing) [inline com.unity.render-pipelines.universal/Editor/2D/Light2DEditor.cs:646]: Combine the two if conditions?

File: com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs [hunk 2]
```diff
@@ -1,11 +1,22 @@
...
 namespace UnityEngine.Experimental.Rendering.Universal
 {
     internal static class LightUtility
     {
+        // public static bool CheckForChange<T>(T a, ref T b) where T : struct
+        // {
+        //     var changed = !EqualityComparer<T>.Default.Equals(a, b);
+        //     b = a;
+        //     return changed;
+        // }
```
Comment (rainsing) [inline com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs:18]: I think we can leave only this generic version and get rid of all other overloads.

File: com.unity.render-pipelines.universal/Runtime/2D/Renderer2DData.cs [hunk 3]
```diff
@@ -76,8 +63,8 @@ public enum Renderer2DDefaultMaterialType
         [SerializeField, Reload("Runtime/Data/PostProcessData.asset")]
         PostProcessData m_PostProcessData = null;
 
-        public float hdrEmulationScale => m_HDREmulationScale;
-        public Light2DBlendStyle[] lightBlendStyles => m_LightBlendStyles;
+        internal float hdrEmulationScale => m_HDREmulationScale;
+        internal Light2DBlendStyle[] lightBlendStyles => m_LightBlendStyles;
```
Comment (rainsing) [inline com.unity.render-pipelines.universal/Runtime/2D/Renderer2DData.cs:67]: These changes may fail the package validation tests.

File: com.unity.render-pipelines.universal/Runtime/2D/Passes/Utility/RendererLighting.cs [hunk 4]
```diff
@@ -586,33 +585,36 @@ private static Material CreateLightMaterial(Renderer2DData rendererData, Light2D
...
 
-            if (rendererData.lightMaterials[materialIndex] == null)
-                rendererData.lightMaterials[materialIndex] = CreateLightMaterial(rendererData, light, isVolume);
+            if (!rendererData.lightMaterials.TryGetValue(materialIndex, out var material))
+            {
+                material = CreateLightMaterial(rendererData, light, isVolume);
+                rendererData.lightMaterials[materialIndex] = material;
+            }
 
             return rendererData.lightMaterials[materialIndex];
```
Comment (rainsing) [inline com.unity.render-pipelines.universal/Runtime/2D/Passes/Utility/RendererLighting.cs:596]: We could use the output of `TryGetValue` above instead of doing a lookup again here.

Comment (UnityPaul) [reply com.unity.render-pipelines.universal/Runtime/2D/Passes/Utility/RendererLighting.cs:596]: fixed

File: com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs (hunk 2 above)
Comment (UnityPaul) [reply com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs]: i'm not sure I understand the performance characteristics of EqualityComparer

File: com.unity.render-pipelines.universal/Runtime/2D/Renderer2DData.cs (hunk 3 above)
Comment (UnityPaul) [reply com.unity.render-pipelines.universal/Runtime/2D/Renderer2DData.cs:67]: how do we remove API?

Comment (rainsing) [reply com.unity.render-pipelines.universal/Runtime/2D/Renderer2DData.cs:67]: I think we need to first mark them as obsolete, then remove them in a later major version.

Comment (rainsing) [reply com.unity.render-pipelines.universal/Runtime/2D/Renderer2DData.cs]: I've reverted this change for now. We will do a holistic review of public APIs before we get out of preview.

File: com.unity.render-pipelines.universal/Runtime/2D/Light2DShape.cs [hunk 5]
```diff
@@ -34,26 +29,13 @@ sealed public partial class Light2D : MonoBehaviour
...
         //==========================================================================================
 
-        internal bool IsShapeLight()
-        {
-            return m_LightType != LightType.Point;
-        }
+        internal bool isShapeLight => m_LightType != LightType.Point;
 
-        BoundingSphere GetShapeLightBoundingSphere()
+        float GetShapeLightBoundingSphere(Bounds bounds)
```
Comment (UnityPaul) [inline com.unity.render-pipelines.universal/Runtime/2D/Light2DShape.cs:34]: rename this to `CalculateBoundingSphereRadius`

Comment (rainsing) [reply com.unity.render-pipelines.universal/Runtime/2D/Light2DShape.cs]: Fixed.

File: com.unity.render-pipelines.universal/Runtime/2D/Light2D.cs [hunk 6]
```diff
@@ -172,23 +174,37 @@ internal int GetTopMostLitLayer()
...
-                    bounds = LightUtility.GenerateParametricMesh(m_Mesh, 1.412135f, 0, 0, 4);
+                    m_LocalBounds = LightUtility.GenerateParametricMesh(m_Mesh, 1.412135f, 0, 0, 4);
                     break;
             }
-            boundingSphereRadius = isShapeLight ? CalculateBoundingSphereRadius(bounds) : m_PointLightOuterRadius;
+        }
+
+        private void UpdateBoundingSphere()
+        {
+            if (!isShapeLight)
```
Comment (UnityPaul) [inline com.unity.render-pipelines.universal/Runtime/2D/Light2D.cs:168]: change `isShapeLight` to something that actually what it means... `isPointLight` because everywhere it is used, it is negated...

File: com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs [hunk 7]
```diff
@@ -170,6 +170,7 @@ public static Bounds GenerateSpriteMesh(Mesh mesh, Sprite sprite)
             mesh.SetVertexBufferParams(vertices.Length, SpriteLightMeshVertex.VertexLayout);
             mesh.SetVertexBufferData(vertices, 0, 0, vertices.Length);
             mesh.SetIndices(srcIndices, MeshTopology.Triangles, 0, true);
+            mesh.RecalculateBounds();
```
Comment (UnityPaul) [inline com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs:173]: why is this needed? the `true` in the SetIndices should recalculate the bounds... or does it not?

Comment (rainsing) [reply com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs]: Removed.

File: com.unity.render-pipelines.universal/Runtime/2D/Light2D.cs (hunk 6 above)
Comment (rainsing) [reply com.unity.render-pipelines.universal/Runtime/2D/Light2D.cs]: Fixed.

**Review by rainsing — APPROVED** [review]

**Review by ellioman — APPROVED** [review]

**Comment by ellioman** [comment]
The failing test is because a purchase package is in Hybrid test project's manifest. It has been removed on master so everything is good on Yamato side.

**Comment by foobraco** [comment]
Is this refactor going to be backported to the 7.x.x branch?

**Comment by rainsing** [comment]
> Is this refactor going to be backported to the 7.x.x branch?

No. We don't have a plan to backport this.
